src/logger.c

```diff
@@ -0,0 +1,197 @@
+  /* create a TCP socket */
+  if ((sock = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
+    return;
+  }
+  address.sin_family = AF_INET;
+  address.sin_port = htons(LOGGING_PORT);
+  address.sin_addr.s_addr = INADDR_ANY;
+  if (bind(sock, (struct sockaddr *)&address, sizeof(address)) < 0) {
+    return;
+  }
+  listen(sock, 5);
+}
+
+/**
+ * @brief Set the program log level. Blocks indefinitely until resource is
+ * available.
+ *
+ * @param level target logging level
+ */
+void logger_set_level(enum logger_level level) {
+  if (xSemaphoreTake(mutex_handle, portMAX_DELAY) == pdTRUE) {
+    _level = level;
+    xSemaphoreGive(mutex_handle);
+  }
+}
+
+/**
+ * @brief Get the program log level. Blocks indefintely until resource is
+ * available.
+ *
+ * @return enum Level
+ */
+enum logger_level logger_get_level(void) {
+  enum logger_level level;
+  if (xSemaphoreTake(mutex_handle, portMAX_DELAY) == pdTRUE) {
+    level = _level;
+    xSemaphoreGive(mutex_handle);
+  }
+  return level;
+}
+
+/**
+ * @brief Write logging message to buffer.
+ *
+ * @param level log level enum
+ * @param func macro expanded log containing function name
+ * @param line macro expanded log line
+ * @param fmt log string formatter
+ * @param ... variable arguments for string formatter
+ */
+void logger_out(const enum logger_level level, const char *func, const int line, const char *fmt, ...) {
+  log_t log;
+  va_list args;
+  // filter output by log level
+  if (logger_get_level() > level) {
+    return;
+  }
+  log.epoch = xTaskGetTickCount();
+  log.level = level;
+  log.line = line;
+  memcpy(log.func, func, FUNC_NAME_MAX_LEN);
+  // variable arguments must be processed here otherwise they will go out of scope
+  va_start(args, fmt);
+  // TODO: implement buffer overflow handler (add truncating)
+  vsnprintf(log.message, MAX_LOGGING_LINE_LEN, fmt, args);
+  va_end(args);
+  // write log struct by reference to queue, block for 20 ticks if queue is full
+  xQueueSendToBack(queue_handle, (void *)&log, 20);
+}
+
+void logger_task(void *pv_params) {
+  log_t log;
+  int client_fd;
+  while (1) {
+    client_fd = accept(sock, (struct sockaddr *)&remotehost, (socklen_t *)&size);
+    while (1) {
+      xQueueReceive(queue_handle, &log, portMAX_DELAY);
+      if (client_fd < 0) {
+        continue;
+      }
+      build_log_string(&log);
+      ssize_t bytes_sent = send(client_fd, &log.message, strlen(log.message), 0);
+      /* Check for errors or client disconnect */
+      if (bytes_sent <= 0) {
+        /* Connection closed by client */
+        close(client_fd);
+        break;
+      }
+    }
+  }
+}
+
+static const char *_get_level_str(enum logger_level level) {
+  switch (level) {
+    case LOGGER_TRACE:
+      return "TRACE";
+    case LOGGER_INFO:
+      return "INFO";
+    case LOGGER_WARNING:
+      return "WARNING";
+    case LOGGER_ERROR:
+      return "ERROR";
+    default:
+      return "CRITICAL";
+  }
+}
+
+static void build_log_string(log_t *log) {
```

Comment:
@BrandonLucasZhu can you add a TODO comment just to indicate that there is a bug in this implementation. This needs to be addressed by someone in a future ticket.